interfaces/builtin/mount_control.go

```diff
@@ -345,15 +345,21 @@ func validateMountTypes(types []string) error {
 	return nil
 }
 
-func validateMountOptions(options []string) error {
-	if len(options) == 0 {
+func validateMountOptions(mountInfo *MountInfo) error {
+	if len(mountInfo.options) == 0 {
 		return errors.New(`mount-control "options" cannot be empty`)
 	}
-	for _, o := range options {
-		if !strutil.ListContains(allowedMountOptions, o) {
-			return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+	if !(len(mountInfo.types) == 1 && mountInfo.types[0] == "functionfs") {
```

Comment:
It seems we still want to make sure no apparmor special chars are used (like we do for "what")

(super-nitpick) sorry, I am a a bit annoying here, somehow something triggered some refactor impulse in me and I couldn't stop myself - hope it's still somewhat useful. 

Given that we already have code that does the "len(mountInfo.types == 1 ..." I would extract a helper like:
```go
func (mi *MountInfo) isType(typ string) bool {
	return len(mi.types()) == 1 && mi.types[0] == typ
}
```
and use it here like:

```go
if mountInfo.isType("functionfs") {
    return validateNoApparmorRegexpWithError("cannot use mount-control "options": %w", err)
}
for _, o := range mountInfo.options {
```
and define a validateNoApparmorRegexpWithError() that is also used in "validateWhatAttr" as something like:
```go
func validateNoApparmorRegexpWithError(errPrefix string, sl ...string) error {
	for _, s := range sl {
		if err := apparmor_sandbox.ValidateNoAppArmorRegexp(s); err != nil {
			return fmt.Errorf(errPrefix+`: %w`, err)
		}
	}
	return nil
}
```
full diff with my suggestion:  https://paste.ubuntu.com/p/WF9TbtbQXw/

Reply:
Great suggestions thanks.

Some things to double-check:
- In addition to `useType()` I propose introducing `hasType()` to replace repeated (2) instances of `len(mi.types) > 0`
- There is a subtlety with moving the `optionIncompatibleWithFsType` check into `validateMountOptions`: It used to check that bind is not used with any fs type. As proposed above this logic is modified, which will allow functionfs to specify bind, seemingly breaking the rule and intention? I proposed slightly modified logic to still run this check for functionfs which will result in error if bind is used with functionfs. Also added case to unit test to bring describe this case.

Thoughts?
https://paste.ubuntu.com/p/mmYBSk77s8/

Reply:
Thanks! (1) looks fine as for (2) I was aware but did not look deep enough. Thanks for flagging this, we need to keep checking for "bind" and disallow it. The issue is that the mount command will happily accept: `mount -o bind -t functionfs /etc/fstab /tmp/xxx` even if the combination does not make any sense. The "-o bind" wins here so we need to ensure it's not used.